numpy/lib/function_base.py

```diff
@@ -175,6 +175,16 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
             raise AttributeError(
                 'max must be larger than min in range parameter.')
 
+    # Histogram is an integer or a float array depending on the weights.
+    if weights is None:
+        ntype = np.dtype(np.intp)
+    else:
+        ntype = weights.dtype
+
+    # We set a block size, as this allows us to iterate over chunks when
+    # computing histograms, to minimize memory usage.
+    BLOCK = 65536
```

Comment:
Just to clarify, I just used the value that was set for the original slow method, but as I indicated in https://github.com/numpy/numpy/pull/6100#issuecomment-123618622 for large arrays we'd actually benefit from a slightly larger block size.

Rather than spend too much time trying to find a one-size-fits-all block size, I would suggest leaving the block size for now since this is code that I plan to try and replace with a C iteration (as @jaimefrio and I were discussing).

However, I agree that the keep operation could be done in-place instead. I'll look at this now.